compiler/modules/front.syntax/tests/canonicalTree.bal

```diff
@@ -0,0 +1,295 @@
+                         normalizeExprSyntaxNode(childNodes[2], true),
+                         normalizeSyntaxNode(childNodes[3])];
+    }
+    else if astNode is FieldAccessLExpr {
+        newChildNodes = [normalizeLExprSyntaxNode(childNodes[0])];
+        // JBUG can't use query expression
+        // newChildNodes.push(...from int i in 1 ..< childNodes.length() select normalizeSyntaxNode(childNodes[i]));
+        foreach int i in 1 ..< childNodes.length() {
+            newChildNodes.push(normalizeSyntaxNode(childNodes[i]));
+        }
+    }
+    else {
+        newChildNodes = from SyntaxNode child in childNodes select normalizeSyntaxNode(child);
+    }
+    return { astNode, childNodes: newChildNodes };
+}
+
+function normalizeTypeDescSyntaxNode(SyntaxNode node, boolean|BinaryTypeOp wrap) returns SyntaxNode {
+    if node is TerminalSyntaxNode {
+        return node;
+    }
+    AstNode astNode = node.astNode;
+    if astNode !is TypeDesc {
+        panic error("expected type desc");
+    }
+    SyntaxNode[] childNodes = node.childNodes;
+    if astNode is UnaryTypeDesc && astNode.op == "(" {
+        return normalizeTypeDescSyntaxNode(childNodes[1], wrap);
+    }
+    SyntaxNode[] newChildNodes;
+    if astNode is BinaryTypeDesc {
+        boolean noWrap = wrap == false || wrap == astNode.op;
+        newChildNodes = conditionallWrapChildNodes(!noWrap, [normalizeTypeDescSyntaxNode(childNodes[0], astNode.op),
+                                                             normalizeSyntaxNode(childNodes[1]),
+                                                             normalizeTypeDescSyntaxNode(childNodes[2], astNode.op)]);
+    }
+    else if astNode is TupleTypeDesc {
+        newChildNodes = conditionallWrapChildNodes(wrap != false, from SyntaxNode child in childNodes select normalizeSyntaxNode(child));
+    }
+    else if astNode is ArrayTypeDesc {
+        newChildNodes = [normalizeTypeDescSyntaxNode(childNodes[0], true)];
+        // JBUG can't use query expression
+        // newChildNodes.push(...from int i in 1 ..< childNodes.length() select normalizeSyntaxNode(childNodes[i]));
+        foreach int i in 1 ..< childNodes.length() {
+            newChildNodes.push(normalizeSyntaxNode(childNodes[i]));
+        }
+        newChildNodes = conditionallWrapChildNodes(wrap != false, newChildNodes);
+    }
+    else {
+        newChildNodes = from SyntaxNode child in childNodes select normalizeSyntaxNode(child);
+    }
+    return { astNode, childNodes: newChildNodes };
+}
+
+function conditionallWrapChildNodes(boolean condition, SyntaxNode[] childNodes) returns SyntaxNode[] {
+    if condition == false {
+        return childNodes;
+    }
+    else {
+        SyntaxNode[] newChildNodes = [{ token: "(" }];
+        newChildNodes.push(...childNodes);
+        newChildNodes.push({ token: ")" });
+        return newChildNodes;
+    }
+}
+
+function syntaxNodeToString(SyntaxNode|RootSyntaxNode node) returns string {
```

Comment:
Put syntaxNodeToString and related functions into syntaxNode. Put the parenthesis normalization/canonicalization into a separate file and make the file naming consistent with function naming.

The recursive functions should append strings to a `string[]`. There should be a top-level function on the root that concatenates this array. (At the moment you are doing lots of concatenations.)

Reply:
fixed